py/ringbuf.c

```diff
@@ -116,5 +103,153 @@ int ringbuf_put_bytes(ringbuf_t *r, const uint8_t *data, size_t data_len) {
     }
     memcpy(r->buf + iput, datap, iput_a - iput);
     r->iput = iput_a;
+}
+
+// Returns:
+//    0: Success
+//   -1: Not enough data available to complete read (try again later)
+//   -2: Requested read is larger than buffer - will never succeed
+int ringbuf_get_bytes(ringbuf_t *r, uint8_t *data, size_t data_len) {
+    if (ringbuf_avail(r) < data_len) {
+        return (r->size <= data_len) ? -2 : -1;
+    }
+    _ringbuf_memcpy_get(r, data, data_len);
     return 0;
 }
+
+// Returns:
+//    0: Success
+//   -1: Not enough free space available to complete write (try again later)
+//   -2: Requested write is larger than buffer - will never succeed
+int ringbuf_put_bytes(ringbuf_t *r, const uint8_t *data, size_t data_len) {
+    if (ringbuf_free(r) < data_len) {
+        return (r->size <= data_len) ? -2 : -1;
+    }
+    _ringbuf_memcpy_put(r, data, data_len);
+    return 0;
+}
+
+#if MICROPY_PY_MICROPYTHON_RINGBUFFER
+#include "py/runtime.h"
+#include "py/stream.h"
+#include "py/mphal.h"
+
+#ifndef MICROPY_PY_MICROPYTHON_RINGBUFFER_USE_MEMCPY
+#define MICROPY_PY_MICROPYTHON_RINGBUFFER_USE_MEMCPY (1)
```

Comment:
I don't think it's worth providing this option to turn memcpy optimisation off.  Just always enable it, ie always use memcpy.  It doesn't cost much extra and is faster (and you want fast when using a ringbuffer).